src/partners/kado.ts

```diff
@@ -0,0 +1,210 @@
+import {
+  asArray,
+  asBoolean,
+  asDate,
+  asMaybe,
+  asNumber,
+  asObject,
+  asOptional,
+  asString,
+  asValue
+} from 'cleaners'
+
+import {
+  asStandardPluginParams,
+  PartnerPlugin,
+  PluginParams,
+  PluginResult,
+  StandardTx,
+  Status
+} from '../types'
+import { datelog, retryFetch, smartIsoDateFromTimestamp, snooze } from '../util'
+
+// Define cleaner for individual transactions in onRamps and offRamps
+const asTxType = asValue('buy', 'sell')
+
+const asTransaction = asObject({
+  _id: asString,
+  walletAddress: asString,
+  createdAt: asDate,
+  type: asTxType,
+  walletType: asString,
+  cryptoCurrency: asString,
+  network: asString
+})
+
+const asOnRampTx = asObject({
+  ...asTransaction.shape,
+  receiveUnitCount: asNumber,
+  paidAmountUsd: asNumber,
+  paymentMethod: asString
+})
+
+const asOffRampTx = asObject({
+  ...asTransaction.shape,
+  depositUnitCount: asNumber,
+  receiveUsd: asNumber
+  // disburseMethod: asString
+})
+
+// Define cleaner for the main data structure
+const asResponse = asObject({
+  success: asBoolean,
+  // message: asString,
+  data: asObject({
+    onRamps: asArray(asOnRampTx),
+    offRamps: asArray(asOffRampTx)
+  })
+})
+
+const MAX_RETRIES = 5
+const QUERY_LOOKBACK = 1000 * 60 * 60 * 24 * 30 // 30 days
+const QUERY_TIME_BLOCK_MS = QUERY_LOOKBACK
+
+// const statusMap: { [key in PartnerStatuses]: Status } = {
+//   DONE: 'complete',
+//   other: 'other'
+// }
+
+export async function queryKado(
+  pluginParams: PluginParams
+): Promise<PluginResult> {
+  const { settings, apiKeys } = asStandardPluginParams(pluginParams)
+  const { apiKey } = apiKeys
+  let { latestIsoDate } = settings
+
+  // API doesn't currently support paging by date but leave this in here
+  // for when it does
+  if (latestIsoDate === '2018-01-01T00:00:00.000Z') {
+    latestIsoDate = new Date('2024-01-01T00:00:00.000Z').toISOString()
+  }
+
+  let lastCheckedTimestamp = new Date(latestIsoDate).getTime() - QUERY_LOOKBACK
+  if (lastCheckedTimestamp < 0) lastCheckedTimestamp = 0
+
+  const ssFormatTxs: StandardTx[] = []
+  let retry = 0
+  let startTime = lastCheckedTimestamp
+
+  // No looping since all txs are returned in one call for now
+  // while (true)
+  {
```

Comment:
Why not just make a separate commit for the while-loop and time variables which can be in a separate wip-branch?